### PR TITLE
FEATURE: Configurable auto-bump cooldown

### PR DESCRIPTION
**Pull request by Drenmi**
Replaces #20038.

### What is this feature?

This addresses the following need:

> We want to add a category-level option to configure time between auto-bumps for a single topic **so once set as needed, auto-bump doesn’t interfere with auto-close unexpectedly.**

Currently the auto-bump cooldown is hard-coded to 24 hours.

---

Below follows a rough step-by-step of the auto-bump procedure for reference:

> Every 15 minutes we look for a single topic across all categories to bump. The selection process is:
> 
> 1. Get all categories that have daily topics to bump > 0.
> 2. Shuffle them.
> 3. Go through them one by one to find the first one that hasn’t met the limit and has at least one bumpable topic.
> 4. Bump the topic in that category which has gone the longest without a bump.
>
> Criteria for determining if a topic is bumpable or not:
>
> - Must be visible.
> - Must be listed.
> - Must NOT be scheduled for auto-bump using a topic timer.
> - **Must NOT have been bumped in the last 24 hours.**
> - Must NOT be pinned.
> - Must NOT be closed.
> - Must NOT be archived.

This change makes the highlighted 24 hours part configurable (defaulting to 24 hours), and the rest of the process remains the same.

### How is this implemented?

This uses the new `CategorySetting` model associated with `Category`. We decided to add this because we want to move away from custom fields due to the lack of type casting and validations, but we want to keep the loading of these optional as they are not needed for almost all of the flows.

Category settings will be back-filled to all categories as part of this change, and creating a new category will now also create a category setting.

### Screenshot from the setting page

Screenshot 2023-03-02 at 4 17 32 PM

### Manual tests performed

- [x] Creating a new category also creates a category setting (with values if provided.)
- [x] Updating an existing category also updates category setting.
- [x] Deleting a category also deletes the category setting.

### CODE REVIEW

File: app/models/category.rb [hunk 1]
```diff
@@ -46,10 +46,22 @@ class Category < ActiveRecord::Base
   has_many :topic_timers, dependent: :destroy
   has_many :upload_references, as: :target, dependent: :destroy
 
-  has_one :category_setting, dependent: :destroy
+  # TODO: This should really be `category_setting`,but that conflicts with
+  #       an association added in `discourse-topic-voting` plugin. We should
+  #       have the plugin extend this core model and then rename this.
+  has_one :category_core_setting, class_name: "CategorySetting", dependent: :destroy
```
Comment (Drenmi) [inline app/models/category.rb:52]: I normally wouldn't endorse this kind of "bag-o'-words" naming, but I don't want to lose the connection to `category_setting` in the plugin.

Comment (tgxworld) [reply app/models/category.rb:52]: I've fixed the plugin so you shouldn't have to do this anymore. 

See https://github.com/discourse/discourse-topic-voting/pull/148

Comment (Drenmi) [reply app/models/category.rb:52]: Better! 🙂

File: app/serializers/category_serializer.rb [hunk 2]
```diff
@@ -30,6 +39,10 @@ def include_reviewable_by_group_name?
     SiteSetting.enable_category_group_moderation? && object.reviewable_by_group_id.present?
   end
 
+  def include_category_setting?
+    object.association(:category_setting).loaded?
+  end
```
Comment (Drenmi) [inline app/serializers/category_serializer.rb:44]: Only included if the category settings have been eager loaded first.

File: db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class BackfillAutoBumpCooldownCategorySetting < ActiveRecord::Migration[7.0]
+  def up
+    execute(<<~SQL)
```
Comment (Drenmi) [inline db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:5]: Tested this locally. I had 33 categories and 5 category settings:

```
SELECT COUNT(*) FROM categories;
#=> 33

SELECT COUNT(*) FROM category_settings;
#=> 5
```

Afterwards I have 33 category_settings:

```
SELECT COUNT(*) FROM category_settings;
#=> 33
```

File: db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class BackfillAutoBumpCooldownCategorySetting < ActiveRecord::Migration[7.0]
```
Comment (tgxworld) [inline db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:3]: Sorry can I trouble you to add a spec here for the review? We don't have to commit it but it'll be nice if we can ensure that the migration is doing what it intends to do with a test. You can use the following template too run the migration within an rspec test.

```
# frozen_string_literal: true

require Rails.root.join(
  "db/migrate/20221201024458_make_channel_slugs_unique_with_index.rb",
)

RSpec.describe MakeChannelSlugsUniqueWithIndex do
  it "does whatever the migration is supposed to do" do
    MakeChannelSlugsUniqueWithIndex.new.up
  end
end
```

Comment (Drenmi) [reply db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:3]: Seems perfectly reasonable to me. 🙂

Comment (Drenmi) [reply db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:3]: @tgxworld [Added](https://github.com/discourse/discourse/pull/20507/files#diff-e72d2fe232e2cd88ae210379fcbfc330cc45960cec15b0553a7db690c3d2319bR5).

File: db/migrate/20230301071240_add_auto_bump_cooldown_to_category_settings.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class AddAutoBumpCooldownToCategorySettings < ActiveRecord::Migration[7.0]
+  def change
+    add_column :category_settings, :auto_bump_cooldown, :integer, default: 1
```
Comment (tgxworld) [inline db/migrate/20230301071240_add_auto_bump_cooldown_to_category_settings.rb:5]:
```suggestion
    add_column :category_settings, :auto_bump_cooldown_days, :integer, default: 1
```

What do you think about including the units for the column in the column name itself? It'll make it alot clearer what the unit of the column is without having to read through the code to figure this out.

Comment (Drenmi) [reply db/migrate/20230301071240_add_auto_bump_cooldown_to_category_settings.rb:5]: I think the ideal solution would actually be to use the `ActiveRecord::Types` API to serialize and deserialize durations, as the unit is actually by right part of the data, not identifiers, but also didn't want to extend the scope too far. 🙂  We can add `days` for now.

File: db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb (hunk 4 above)
Comment (tgxworld) [reply db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:3]: Hmm I ran it locally and it fails for me as well. Maybe try to reset your test DB locally?

Comment (Drenmi) [reply db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:3]: After `RAILS_ENV=test rails db:reset db:migrate` it still passes. 😕 

```
bundle exec rspec spec/migrations/backfill_auto_bump_cooldown_category_setting_spec.rb      

Randomized with seed 35015
-- execute("INSERT INTO\n  category_settings(\n    category_id,\n    auto_bump_cooldown,\n    created_at,\n    updated_at\n  )\nSELECT\n  id,\n  1,\n  NOW(),\n  NOW()\nFROM categories\nON CONFLICT (category_id)\nDO\n  UPDATE SET\n    auto_bump_cooldown = 1,\n    updated_at = NOW();\n")
   -> 0.0004s
.

Finished in 0.26009 seconds (files took 1.65 seconds to load)
1 example, 0 failures
```

**Review by tgxworld — APPROVED** [review]
Changes look good to me :+1:

File: db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb (hunk 4 above)
Comment (Drenmi) [reply db/migrate/20230308042434_backfill_auto_bump_cooldown_category_setting.rb:3]: Okay. Problem was solved by `RAILS_ENV=test rails db:drop db:create db:migrate`.

**Pull request merged**